runtime/src/governance/election.rs

```diff
@@ -181,51 +183,93 @@ impl<T: Trait> Module<T> {
         <VotingPeriod<T>>::put(params.voting_period);
         <RevealingPeriod<T>>::put(params.revealing_period);
-        <CouncilSize<T>>::put(params.council_size);
         <MinCouncilStake<T>>::put(params.min_council_stake);
         <NewTermDuration<T>>::put(params.new_term_duration);
+        <CouncilSize<T>>::put(params.council_size);
         <CandidacyLimitMultiple<T>>::put(params.candidacy_limit_multiple);
     }
 
-    pub fn is_election_running() -> bool {
-        Self::stage().is_some()
-    }
-
+    /// Starts an election. Will fail if an election is already running
+    /// Initializes transferable stakes. Assumes election parameters have already been set.
     fn start_election(current_council: Option<Seats<T::AccountId, T::Balance>>) -> Result {
         ensure!(!Self::is_election_running(), "election already in progress");
+        ensure!(Self::existing_stake_holders().len() == 0, "stake holders must be empty");
+        ensure!(Self::applicants().len() == 0, "applicants must be empty");
+        ensure!(Self::commitments().len() == 0, "commitments must be empty");
 
-        // take snapshot of council and backing stakes of an existing council
+        // Take snapshot of seat and backing stakes of an existing council
+        // Its important to ensure the election system takes ownership of these stakes, and is responsible
+        // to return any unused stake to origin owners.
         current_council.map(|c| Self::initialize_transferable_stakes(c));
 
         Self::move_to_announcing_stage();
         Ok(())
     }
 
-    fn new_period(length: T::BlockNumber) -> T::BlockNumber {
-        <system::Module<T>>::block_number() + length
-    }
+    /// Sets announcing stage. Can be called from any stage and assumes all preparatory work
+    /// for entering the stage has been performed.
+    /// Bumps the election round.
+    fn move_to_announcing_stage() {
+        let next_round = <Round<T>>::mutate(|n| { *n += 1; *n });
+
+        let new_stage_ends_at = Self::current_block_number_plus(Self::announcing_period());
 
-    fn bump_round() -> u32 {
-        <Round<T>>::mutate(|n| {
-            *n += 1;
-            *n
-        })
+        <Stage<T>>::put(ElectionStage::Announcing(new_stage_ends_at));
+
+        Self::deposit_event(RawEvent::AnnouncingStarted(next_round));
     }
 
-    fn move_to_announcing_stage() -> T::BlockNumber {
-        let period = Self::new_period(Self::announcing_period());
+    /// Sets announcing stage. Can be called from any stage and assumes all preparatory work
+    /// for entering the stage has been performed.
+    fn move_to_voting_stage() {
+        let new_stage_ends_at = Self::current_block_number_plus(Self::voting_period());
 
-        <Stage<T>>::put(ElectionStage::Announcing(period));
+        <Stage<T>>::put(ElectionStage::Voting(new_stage_ends_at));
 
-        let next_round = Self::bump_round();
+        Self::deposit_event(RawEvent::VotingStarted());
```

Comment:
Could be helpful to add `new_stage_ends_at ` param to events as well (events: VotingStarted, RevealingStarted)

Reply:
Yes I didn't put much thought into what information would be useful for ui, so feel free to add as needed.